File: strata-js/src/fmt.ts
import { Value } from "./value.js";

export enum FormatOptions {
    PRETTY = "PRETTY",
    AST = "AST",
}

export function fmt(opts: FormatOptions, value: Value): string {
    switch (opts) {
        case FormatOptions.PRETTY: {
            const f = new PrettyFmt();
            f.value(value);
            return f.out;
        }
        case FormatOptions.AST:
            return prettyDebug(value);
    }
}

class PrettyFmt {
    out = "";
    depth = 0;
    indent = 2;

    private pad(): string {
        return " ".repeat(this.depth * this.indent);
    }

    value(v: Value): void {
        switch (v.kind) {
            case "null":
                this.out += "null";
                break;
            case "bool":
                this.out += String(v.value);
                break;
            case "int":
                this.out += String(v.value);
                break;
            case "string":
                this.out += `"${v.value}"`;
                break;
            case "bytes":
                this.bytes(v.value);
                break;
            case "list":
                this.list(v.value as Value[]);
                break;
            case "map":
                this.map(v.value as Map<string, Value>);
                break;
        }
    }

    private bytes(bytes: Uint8Array): void {
        this.out += "[";
        for (const b of bytes) {
            this.out += b.toString(16).padStart(2, "0");
        }
        this.out += "]";
    }

    private list(list: Value[]): void {
        this.out += "[";
        for (let i = 0; i < list.length; i++) {
            this.value(list[i]!);
            if (i + 1 !== list.length) this.out += ", ";
        }
        this.out += "]";
    }

    private map(map: Map<string, Value>): void {
        this.out += "{\n";
        this.depth += 1;

        const entries = [...map.entries()].sort(([a], [b]) => a.localeCompare(b));

        for (const [k, v] of entries) {
            this.out += this.pad();
            this.out += k;
            this.out += ": ";
            this.value(v);
            this.out += "\n";
        }

        this.depth -= 1;
        this.out += this.pad();
        this.out += "}";
    }
}

function prettyDebug(value: unknown): string {
    return JSON.stringify(value, null, 2);
}

File: strata-rs/src/fmt.rs
use std::collections::BTreeMap;

use crate::value::Value;

#[derive(Clone, Copy)]
pub enum FormatOptions {
    PRETTY,
    AST,
}

pub fn fmt(opts: FormatOptions, value: &Value) -> String {
    match opts {
        FormatOptions::PRETTY => {
            let mut f = PrettyFmt {
                out: String::new(),
                depth: 0,
                indent: 2,
            };
            f.value(value);
            f.out
        }
        FormatOptions::AST => format!("{:#?}", value),
    }
}

struct PrettyFmt {
    out: String,
    depth: usize,
    indent: usize,
}

impl PrettyFmt {
    fn pad(&self) -> String {
        " ".repeat(self.depth * self.indent)
    }

    fn value(&mut self, value: &Value) {
        match value {
            Value::Null => self.out.push_str("null"),
            Value::Bool(b) => self.out.push_str(&b.to_string()),
            Value::Int(i) => self.out.push_str(&i.to_string()),
            Value::String(s) => {
                self.out.push('"');
                self.out.push_str(s);
                self.out.push('"')
            }
            Value::Bytes(b) => self.bytes(b),
            Value::List(l) => self.list(l),
            Value::Map(m) => self.map(m),
        }
    }

    fn bytes(&mut self, bytes: &[u8]) {
        self.out.push('[');
        for byte in bytes {
            self.out.push_str(&format!("{:02x}", byte));
        }
        self.out.push(']');
    }

    fn list(&mut self, list: &[Value]) {
        self.out.push('[');

        for (i, v) in list.iter().enumerate() {
            self.value(v);
            if i + 1 != list.len() {
                self.out.push_str(", ");
            }
        }

        self.out.push(']');
    }

    fn map(&mut self, map: &BTreeMap<String, Value>) {
        self.out.push_str("{\n");
        self.depth += 1;

        for (k, v) in map.iter() {
            self.out.push_str(&self.pad());
            self.out.push_str(k);
            self.out.push_str(": ");
            self.value(v);
            self.out.push('\n');
        }

        self.depth -= 1;
        self.out.push_str(&self.pad());
        self.out.push('}');
    }
}
